channeldb/reports.go

```diff
@@ -0,0 +1,332 @@
+	ErrNoChainHashBucket = errors.New("no chain hash bucket")
+
+	// ErrNoChannel is returned when a channel is not found in the chain
+	// hash bucket.
+	ErrNoChannel = errors.New("channel bucket not found")
+)
+
+// ResolverOutcome describes the outcome for the resolver that the contract
+// court reached. This state is not necessarily final, htlcs on our own
+// commitment are resolved across two resolvers.
+type ResolverOutcome uint8
+
+const (
+	resolverOutcomeUnknown ResolverOutcome = iota
+
+	// ResolverOutcomeAnchorRecovered indicates that we swept our own
+	// anchor.
+	ResolverOutcomeAnchorRecovered
+
+	// ResolverOutcomeAnchorLost indicates that we did not sweep our anchor,
+	// or another party claimed it.
+	ResolverOutcomeAnchorLost
+
+	// ResolverOutcomeAnchorUnclaimed indicates that the anchor was left
+	// unspent.
+	ResolverOutcomeAnchorUnclaimed
+
+	// ResolverOutcomeOutgoingHtlcTimeout indicates the we successfully
+	// timed out one of our outgoing htlcs on chain.
+	ResolverOutcomeOutgoingHtlcTimeout
+
+	// ResolverOutcomeOutgoingHtlcClaim indicates that the remote party
+	// claimed one of our outgoing htlcs on chain with the preimage.
+	ResolverOutcomeOutgoingHtlcClaim
+
+	// ResolverOutcomeIncomingHtlcClaimed indicates that we claimed an
+	// incoming htlc with its preimage on chain.
+	ResolverOutcomeIncomingHtlcClaimed
+
+	// ResolverOutcomeIncomingHtlcSuccessTx indicates that we broadcast the
+	// first stage of a two stage htlc claim for an incoming htlc.
+	ResolverOutcomeIncomingHtlcSuccessTx
+
+	// ResolverOutcomeIncomingHtlcTimedOut indicates that an incoming htlc
+	// timed out on chain, so we stopped waiting for the preimage.
+	ResolverOutcomeIncomingHtlcTimedOut
+
+	// ResolverOutcomeIncomingHtlcFailed indicates that our peer tried to
+	// settle an incoming htlc on chain, but we failed their attempt.
+	ResolverOutcomeIncomingHtlcFailed
+
+	// ResolverTypeOutgoingHtlcTimeoutTx indicates that we broadcast the
+	// first stage of a two stage htlc timeout for an outgoing htlc.
+	ResolverTypeOutgoingHtlcTimeoutTx
+
+	// ResolverOutputCommitOutputDelay indicates that we swept our time
+	// locked commitment output on chain.
+	ResolverOutputCommitOutputDelay
+)
+
+// ResolverReport provides an account of the outcome of a resolver. This differs
+// from a ContractReport because it does not necessarily fully resolve the
+// contract; each step of two stage htlc resolution is included.
+type ResolverReport struct {
+	// OutPoint is the outpoint that was spent by this resolver.
+	OutPoint wire.OutPoint
+
+	// Amount is the amount in satoshis that was claimed by this resolution.
+	Amount btcutil.Amount
+
+	// ResolverOutcome indicates the type of resolution that occurred.
+	ResolverOutcome
+
+	// SpendTxID is the transaction ID of the spending transaction that
+	// claimed the outpoint. This may be a sweep transaction, or a first
+	// stage success/timeout transaction.
+	SpendTxID *chainhash.Hash
+}
```

Comment:
Will the user be able to backtrace - in case of htlcs - which htlc this was exactly? Wondering what the user flow looks like. A payment hash and/or htlc index could be useful.

Reply:
With the plan to expose this on `ClosedChannels`, this would be more concerned with getting an account of what fees you paid resolving the channel. We'd need to persist the htlcIndex here as well, so that it can be linked to a payment/invoice.